### PR TITLE
feat: add module api diff tool (SDKCF-4025)

### PR DESCRIPTION
**Pull request by imaginaris**
Added a script for determining API differences between SDK versions (.framework file)
I have a feeling that there might be a better naming for that script file 😊

### CODE REVIEW

File: scripts/sdk-api-diff.sh [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+fi
+if [ ! -d $2 ]; then
+    echo "Invalid path: $2"
+    exit 1
+fi
+if [ ! -d $3 ]; then
+    echo "Invalid path: $3"
+    exit 1
+fi
+REQUEST_TEMPALTE="sdk-api-diff-request-tempalte.yml"
```
Comment (apparition47) [inline scripts/sdk-api-diff.sh]: >template

**Review by apparition47 — CHANGES_REQUESTED** [review]

File: scripts/sdk-api-diff-request-tempalte.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+key.request: source.request.editor.open.interface
+key.name: "5F63C5B8-6D92-44FF-8012-DCA7D787D243"
```
Comment (vincentdelmundo) [inline scripts/sdk-api-diff-request-tempalte.yml]: I think this should be in env var or at least sample value and user will have to update.

**Review by vincentdelmundo — CHANGES_REQUESTED** [review]
Looks good. Just 1 comment/question.

File: scripts/sdk-api-diff.sh [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+set -e
+
+usage="--------------------------------------------
+Description: Compares API differences between two module versions.
+
+Usage: sdk-api-diff <module name> <module A path> <module B path> [OPTIONS]
+
+Options:
+  --target: Target platform and version e.g. x86_64-apple-ios14.0-simulator or arm64-apple-ios14.0
+    (Default: x86_64-apple-ios14.0-simulator)
```
Comment (donnie-jp) [inline scripts/sdk-api-diff.sh]: Is it a good idea to default to the simulator target?
Maybe it doesn't matter.

File: scripts/sdk-api-diff.sh [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+if [ ! -d $2 ]; then
+    echo "Invalid path: $2"
+    exit 1
+fi
+if [ ! -d $3 ]; then
+    echo "Invalid path: $3"
+    exit 1
+fi
+REQUEST_TEMPALTE="sdk-api-diff-request-tempalte.yml"
+if [ ! -f $REQUEST_TEMPALTE ]; then
```
Comment (donnie-jp) [inline scripts/sdk-api-diff.sh]: TEMPALTE -> TEMPLATE
(plus on message below)

File: scripts/sdk-api-diff.sh [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    echo "Missing $REQUEST_TEMPALTE file"
+    exit 1
+fi
+
+hash jq 2> /dev/null || { echo >&2 "ERROR: jq must be installed"; exit 1; }
```
Comment (donnie-jp) [inline scripts/sdk-api-diff.sh]: Maybe provide the install command too? Is it on homebrew?

File: scripts/sdk-api-diff.sh [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+hash sourcekitten 2> /dev/null || { echo >&2 "ERROR: sourcekitten must be installed"; exit 1; }
```
Comment (donnie-jp) [inline scripts/sdk-api-diff.sh]: Maybe add "sourcekitten comes with Xcode so ensure you have a recent version of Xcode installed on your system"

File: scripts/sdk-api-diff.sh [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+MODULE_B_DIR=$3
+
+# Export vars for yaml template
+export TARGET_NAME
+export SDK_DIR
+export MODULE_NAME
+export MODULE_DIR=$MODULE_A_DIR
+
+# Replace the {{KEY}} in yaml file with the values from the environment variables
+cat $REQUEST_TEMPALTE | perl -pe 's/\{\{(\w+)\}\}/$ENV{$1}/eg' > requestA.yml
```
Comment (donnie-jp) [inline scripts/sdk-api-diff.sh]: TEMPALTE -> TEMPLATE

File: scripts/sdk-api-diff.sh [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+export MODULE_DIR=$MODULE_B_DIR
+cat $REQUEST_TEMPALTE | perl -pe 's/\{\{(\w+)\}\}/$ENV{$1}/eg' > requestB.yml
```
Comment (donnie-jp) [inline scripts/sdk-api-diff.sh]: TEMPALTE -> TEMPLATE

Comment (donnie-jp) [reply scripts/sdk-api-diff.sh]: Maybe add an explanatory comment for the regex

**Review by donnie-jp — CHANGES_REQUESTED** [review]
This looks great. Can't wait to try it out 👍 

Please add a README usage guide starting from the step to clone the repo. 
Maybe it's better to give it its own folder like `scripts/sdk-api-diff` for the source and README?

**Comment by donnie-jp** [comment]
> Added a script for determining API differences between SDK versions (.framework file)
> I have a feeling that there might be a better naming for that script file 😊

oh, so it needs a `.framework` and we can only use this for SDKs we package as frameworks?

File: scripts/sdk-api-diff-request-tempalte.yml (hunk 2 above)
Comment (imaginaris) [reply scripts/sdk-api-diff-request-tempalte.yml]: But why update that value? It's just a dummy value

File: scripts/sdk-api-diff.sh (hunk 3 above)
Comment (imaginaris) [reply scripts/sdk-api-diff.sh]: I was thinking that we mostly use x86_64 builds for unit tests etc. I found it the most convenient setting for now but we can also remove that default value and make that parameter required

File: scripts/sdk-api-diff.sh (hunk 5 above)
Comment (imaginaris) [reply scripts/sdk-api-diff.sh]: Yes, homebrew. Good idea

File: scripts/sdk-api-diff.sh (hunk 6 above)
Comment (apparition47) [reply scripts/sdk-api-diff.sh]: Sourcekitten comes with Xcode???

Comment (imaginaris) [reply scripts/sdk-api-diff.sh]: Sourcekitten is an external dependency installed by homebrew (or as a part of jazzy gem) but yeah, I'll add more infor about installation

File: scripts/sdk-api-diff.sh (hunk 7 above)
Comment (apparition47) [reply scripts/sdk-api-diff.sh]: I can appreciate the consistency 😎

**Comment by imaginaris** [comment]
> > Added a script for determining API differences between SDK versions (.framework file)
> > I have a feeling that there might be a better naming for that script file 😊
> 
> oh, so it needs a `.framework` and we can only use this for SDKs we package as frameworks?

Yeah, it has to be a compiled source. It doesn't even work with .xcframework - you need to specify the path to .framework file inside .xcframewok . I can try to make this work with Swift Packages as an improvement

File: scripts/sdk-api-diff-request-tempalte.yml (hunk 2 above)
Comment (vincentdelmundo) [reply scripts/sdk-api-diff-request-tempalte.yml]: ok, maybe we can use "XXXXXX..." or something like "your-id..."

File: scripts/sdk-api-diff.sh (hunk 8 above)
Comment (imaginaris) [reply scripts/sdk-api-diff.sh]: There's a comment for identical regex that's above. Should it be more detailed?

File: scripts/sdk-api-diff.sh (hunk 3 above)
Comment (apparition47) [reply scripts/sdk-api-diff.sh]: It seems like a sensible default to me

File: scripts/sdk-api-diff.sh (hunk 6 above)
Comment (donnie-jp) [reply scripts/sdk-api-diff.sh]: oh, sorry, I was getting this confused with "something" 😕

**Review by donnie-jp — APPROVED** [review]
Awesome 👍

**Comment by donnie-jp** [comment]
> > oh, so it needs a `.framework` and we can only use this for SDKs we package as frameworks?
> 
> Yeah, it has to be a compiled source. It doesn't even work with .xcframework - you need to specify the path to .framework file inside .xcframewok . I can try to make this work with Swift Packages as an improvement

😿 
That's a shame.

**Comment by imaginaris** [comment]
> > > oh, so it needs a `.framework` and we can only use this for SDKs we package as frameworks?
> > 
> > 
> > Yeah, it has to be a compiled source. It doesn't even work with .xcframework - you need to specify the path to .framework file inside .xcframewok . I can try to make this work with Swift Packages as an improvement
> 
> 😿
> That's a shame.

I made it work with SPM projects without the need of building (or having xcodeproj file).
The `.framework` method still needs manual build but it can be automated in the future if need be.

File: scripts/sdk-api-diff-request-tempalte.yml (hunk 2 above)
Comment (imaginaris) [reply scripts/sdk-api-diff-request-tempalte.yml]: Changed to env var

**Review by chloengo — COMMENTED** [review]
Hi

**Pull request merged**